sympy/printing/pycode.py

```diff
@@ -516,6 +516,25 @@ def pycode(expr, **settings):
 }
 
 
+def _unpack_integral_limits(integral_expr):
+    """ helper function for _print_Integral that
+        - accepts an Integral expression
+        - returns a tuple of
+           - a list variables of integration
+           - a list of tuples of the upper and lower limits of integration
+    """
+    integration_vars = []
+    limits = []
+    for integration_range in integral_expr.args[1:]:
+        try:
+            integration_var, lower_limit, upper_limit = integration_range
+        except (TypeError, ValueError):
+            raise NotImplementedError("Only definite integrals are supported")
```

Comment:
Why is there a need to catch both `TypeError` and `ValueError`?

This example is probably fine but in general I would prefer to eliminate exception catching as much as possible from the codebase because it is too often done badly. This can just be:
```python
        if len(integration_range) != 3:
            raise NotImplementedError("Only definite integrals are supported")
        integration_var, lower_limit, upper_limit = integration_range
```
I think that these kinds of checks are very common in code working with integrals. Maybe we should add an `Integral.is_definite_integral` property/method.

Reply:
a TypeError will occur trying to unpack a symbol in the indefinite case but a ValueError will occur unpacking a 2-length tuple in the indefinite integral evaluated at a point case. I looked at adding an `is_definite_integral` property like:
```
index d70867293b..98b961811a 100644
--- a/sympy/integrals/integrals.py
+++ b/sympy/integrals/integrals.py
@@ -90,6 +90,35 @@ def __new__(cls, function, *symbols, **assumptions):
     def __getnewargs__(self):
         return (self.function,) + tuple([tuple(xab) for xab in self.limits])
 
+    @property
+    def is_definite_integral(self):
+        """
+        Returns a boolean indicating whether the Integral is definite
+
+        Examples
+        ========
+
+        >>> Integral(x**2, (x, 0, 1)).is_definite_integral
+        True
+
+        >>> Integral(x**2, x)).is_definite_integral
+        False
+
+        A definite integral of an indefinite integral is a definite integral:
+
+        >>> Integral(x**2, x, (x, 0, 1))).is_definite_integral
+        True
+
+        An indefinite integral of a definite integral is not a definite integral:
+
+        >>> from sympy import Function
+        >>> f = Function("f")
+        >>> Integral(f(x), (x, 0, 1), x).is_definite_integral
+        False
+
+        """
+        return len(self.limits[-1]) == 3
+
     @property
     def free_symbols(self):
         """

```
but I realize I need to know that all the limits in a potentially nested integral are definite. However, using the `limits` property (which wraps the integration variable of an indefinite integral in a 1-arg tuple as opposed to `args[:1]`), I think it's reasonable to iterate over that with a length check instead of the don't-ask-permission approach.